### PR TITLE
Fix precheck issue in substitute out of memory.

### PR DESCRIPTION
**Pull request by nguhoa**
Fix issue: https://github.com/microsoft/Power-Fx/issues/1610
- Fix logic to throw overflow exception if result length of substitute is more than int.MaxValue.
- Check if result length is more than int.MaxValue (StringBuilder capacity as in https://learn.microsoft.com/en-us/dotnet/api/system.text.stringbuilder.maxcapacity?view=net-6.0), then return runtime error in substitute precheck.
- Added tests for 2 cases above.

### CODE REVIEW

**Comment by LucGenetier** [comment]
✅ No public API change.

File: src/tests/Microsoft.PowerFx.Interpreter.Tests/SandboxTests.cs [hunk 1]
```diff
@@ -180,10 +196,16 @@ public async Task InfiniteLoop(string expr)
...
+        [InlineData(612220032, 1, 18, true, 11019960576)]
+        [InlineData(int.MaxValue, 1, 100, true, 214748364700)]
+        [InlineData(100, 10, int.MaxValue, true, 21474836470)]
+        [InlineData(612220032, 0, 18, true, 612220032)]
+        [InlineData(0, 1, int.MaxValue, true, 0)]
+        [InlineData(5, 3, 0, true, 5)]
         [InlineData(4, 1, 5, false, 8)] // just first,  (4-1+5)
         [InlineData(4, 5, 3, false, 4)]
-        public void TestSubstitutePrePoll(int sourceLen, int matchLen, int replacementLen, bool replaceAll, int expectedMaxLenChars)
+        public void TestSubstitutePrePoll(int sourceLen, int matchLen, int replacementLen, bool replaceAll, double expectedMaxLenChars)
```
Comment (MikeStall) [inline src/tests/Microsoft.PowerFx.Interpreter.Tests/SandboxTests.cs:208]: >double [](http://example.com/codeflow?start=108&length=6)

int64?  (still a whole number)

File: src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryText.cs [hunk 2]
```diff
@@ -929,9 +930,9 @@ public static FormulaValue Split(IRContext irContext, StringValue[] args)
         }
 
         // This is static analysis before actually executing, so just use string lengths and avoid contents. 
-        internal static int SubstituteGetResultLength(int sourceLen, int matchLen, int replacementLen, bool replaceAll)
+        internal static double SubstituteGetResultLength(int sourceLen, int matchLen, int replacementLen, bool replaceAll)
```
Comment (MikeStall) [inline src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryText.cs:933]: >double [](http://example.com/codeflow?start=24&length=6)

int64

Comment (MikeStall) [inline src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryText.cs:933]: >SubstituteGetResultLength [](http://example.com/codeflow?start=31&length=25)

For overflow,  should this be a checked context? 


We should still catch the overflow exception

[https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/statements/checked-and-unchecked](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/statements/checked-and-unchecked)

File: src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryText.cs [hunk 3]
```diff
@@ -988,7 +989,13 @@ private static FormulaValue Substitute(EvalVisitor runner, EvalVisitorContext co
             var matchLen = match.Value.Length;
             var replacementLen = replacement.Value.Length;
 
-            var maxLenChars = SubstituteGetResultLength(sourceLen, matchLen, replacementLen, instanceNum < 0);
+            var maxLen = SubstituteGetResultLength(sourceLen, matchLen, replacementLen, instanceNum < 0);
+
+            if (!TryGetIntFromDouble(maxLen, out int maxLenChars))
+            {
+                throw new GovernorException($"Memory overuse: Used {maxLen} characters vs. {int.MaxValue} characters allowed.");
+            }
```
Comment (MikeStall) [inline src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryText.cs:997]: we could catch overflow here and return a runtime error. 

CommonErrors.OverflowError

File: src/tests/Microsoft.PowerFx.Interpreter.Tests/SandboxTests.cs (hunk 1 above)
Comment (nguhoa) [reply src/tests/Microsoft.PowerFx.Interpreter.Tests/SandboxTests.cs:208]: keep this as int.

File: src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryText.cs (hunk 2 above)
Comment (nguhoa) [reply src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryText.cs]: keep int.

**Comment by LucGenetier** [comment]
✅ No public API change.

File: src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryText.cs (hunk 2 above)
Comment (nguhoa) [reply src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryText.cs]: add check overflow.

File: src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryText.cs [hunk 4]
```diff
@@ -933,34 +933,37 @@ internal static int SubstituteGetResultLength(int sourceLen, int matchLen, int r
...
+                    }
                 }
             }
-            else
+            catch (OverflowException e)
             {
-                // Only replace 1 instance 
-                maxLenChars = sourceLen - matchLen + replacementLen;
+                throw e;
             }
```
Comment (MikeStall) [inline src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryText.cs:965]: Do we need this?

File: src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryText.cs [hunk 5]
```diff
@@ -987,8 +990,17 @@ private static FormulaValue Substitute(EvalVisitor runner, EvalVisitorContext co
             var sourceLen = source.Value.Length;
             var matchLen = match.Value.Length;
             var replacementLen = replacement.Value.Length;
+            var maxLenChars = sourceLen;
+
+            try
+            {
+                maxLenChars = SubstituteGetResultLength(sourceLen, matchLen, replacementLen, instanceNum < 0);
+            }
+            catch
```
Comment (MikeStall) [inline src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryText.cs:999]: >catch [](http://example.com/codeflow?start=12&length=5)

This should be: 
catch (OverflowException)

File: src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryText.cs (hunk 4 above)
Comment (nguhoa) [reply src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryText.cs:965]: removed

File: src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryText.cs (hunk 5 above)
Comment (nguhoa) [reply src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryText.cs:999]: updated

**Comment by LucGenetier** [comment]
✅ No public API change.

**Review by MikeStall — APPROVED** [review]
:shipit:

**Comment by LucGenetier** [comment]
✅ No public API change.

**Pull request merged**